mcorpc/authz_rego.go

```diff
@@ -0,0 +1,161 @@
+package mcorpc
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+
+	"github.com/choria-io/go-choria/choria"
+	"github.com/choria-io/go-config"
+	"github.com/open-policy-agent/opa/ast"
+	"github.com/open-policy-agent/opa/rego"
+	"github.com/sirupsen/logrus"
+)
+
+type regoMatcher interface {
+	Set(caller string, actions string, facts string, classes string, groups map[string][]string)
+	SetFile(f string)
+}
+
+type regoPolicy struct {
+	cfg     *config.Config
+	req     *Request
+	agent   *Agent
+	log     *logrus.Entry
+	matcher regoMatcher
+	groups  map[string][]string
+}
+
+type regoPolicyPolicy struct {
+	caller  string
+	actions string
+	facts   string
+	classes string
+	groups  map[string][]string
+	log     *logrus.Entry
+	file    string
+}
+
+func regoPolicyAuthorize(req *Request, agent *Agent, log *logrus.Entry) (bool, error) {
+	logger := log.WithFields(logrus.Fields{
+		"authorizer": "regoPolicy",
+		"agent":      agent.Name(),
+		"request":    req.RequestID,
+	})
+
+	authz := &regoPolicy{
+		cfg:     agent.Config,
+		req:     req,
+		agent:   agent,
+		matcher: &regoPolicyPolicy{log: logger},
+		groups:  make(map[string][]string),
+		log:     logger,
+	}
+
+	log.Infof("Agent is %s", agent.Name())
+
+	return authz.authorize()
+}
+
+func (r *regoPolicy) authorize() (bool, error) {
+	policyFile, err := r.lookupPolicyFile()
+	if err != nil {
+		return false, err
+	}
+
+	pol, err := ioutil.ReadFile(policyFile)
+	if err != nil {
+		return false, err
+	}
+
+	module := string(pol)
+
+	query := rego.New(
+		rego.Query("data.choria.authpolicy.allow"),
+		rego.Module("//choria/policies/"+r.agent.meta.Name, module),
+		rego.Input(r.regoInputs()),
+	)
+
+	rs, err := query.Eval(context.Background())
+
+	switch err := err.(type) {
+	case nil:
+		break
+	case ast.Errors:
+		for _, e := range err {
+			r.log.Info("code: ", e.Code)
+			r.log.Info("row: ", e.Location.Row)
+			r.log.Info("filename: ", policyFile)
+		}
+	default:
+		return false, err
+	}
+
+	return rs[0].Expressions[0].Value.(bool), nil
+}
+
+func (r *regoPolicy) defaultPolicyFilenName() string {
+	return r.cfg.Option("plugin.regopolicy.default_name", "default")
+}
+
+func (r *regoPolicyPolicy) Set(caller string, actions string, facts string, classes string, groups map[string][]string) {
+	r.caller = caller
+	r.actions = actions
+	r.facts = facts
+	r.classes = classes
+	r.groups = groups
+}
+
+func (r *regoPolicyPolicy) SetFile(f string) {
+	r.file = f
+}
+
+func (r *regoPolicy) shouldUseDefault() bool {
+	enabled, err := choria.StrToBool(r.cfg.Option("plugin.regopolicy.enable_default", "n"))
```

Comment:
all these we'll make full config settings so things like bool casting happens automatically